Add '@' button for inserting mentions on mobile

mchowning opened the pull request:
## Description
This PR adds an `@` button for inserting mentions on mobile (https://github.com/wordpress-mobile/gutenberg-mobile/issues/331).

## How has this been tested?
1. Open any `RichText`-based block.
2. Tap the `@` button on the far right of the format toolbar
3. Use the Add Mention UI to select a mention to use.

## Checklist:
- [X] My code is tested.
- [X] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [X] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [X] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [X] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [X] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

mchowning committed:
Add '@' button for inserting mentions on mobile

SergioEstevao committed:
Merge branch 'master' into rnmobile/issue_331/mention_support

SergioEstevao commented on packages/rich-text/src/component/index.native.js:
```diff
@@ -26,13 +30,15 @@ import FormatEdit from './format-edit';
 import { isCollapsed } from '../is-collapsed';
 import { remove } from '../remove';
 import styles from './style.scss';
+import { addMention } from 'react-native-gutenberg-bridge';
```
This import should be on the external dependencies section above.

mchowning replied:
👍Done

mchowning committed:
Move react-native-azted dependency to external dependencies section

ceyhun commented on packages/rich-text/src/component/index.native.js:
```diff
@@ -860,15 +873,31 @@ export class RichText extends Component {
+											record,
+											`@${ mentionUserId }`
+										);
+									} );
```
I'm not sure if it's very important but I'm getting a warning on iOS the first time I cancel the mention selection in the console saying `Possible Unhandled Promise Rejection` I think because it's not handled here :)

SergioEstevao replied:
Good catch I now added an empty handler to remove this warning.

SergioEstevao committed:
Merge branch 'master' into rnmobile/issue_331/mention_support

SergioEstevao committed:
Handle promise rejection.

SergioEstevao committed:
Fix focus issue.

SergioEstevao committed:
Add space after mention

SergioEstevao committed:
Update selection onFocus.

SergioEstevao committed:
Merge branch 'master' into rnmobile/issue_331/mention_support

SergioEstevao committed:
Check for site capabilities for mentions support

SergioEstevao committed:
Merge branch 'rnmobile/release-1.27.0' into rnmobile/issue_331/mention_support

SergioEstevao committed:
Add the mention button inside a toolbar.

SergioEstevao committed:
Use HOC for site capabilities.

SergioEstevao committed:
Merge branch 'master' into rnmobile/issue_331/mention_support

mchowning committed:
Only include space after mention on iOS

On Android we are discarding selections when we think they might get
stripped by Aztec, so adding the space to the ends results in the
selection values getting discarded.

SergioEstevao committed:
Merge branch 'master' into rnmobile/issue_331/mention_support

# Conflicts:
#	packages/components/src/index.native.js

cameronvoell commented on packages/rich-text/src/component/index.native.js:
```diff
@@ -408,9 +427,8 @@ export class RichText extends Component {
-		const index = undefined;
 
-		onSelectionChange( index, index );
+		onSelectionChange( this.selectionStart, this.selectionEnd );
```
Curious if this is related to mentions or if there is a specific case that this update is addressing? @SergioEstevao

SergioEstevao replied:
When doing insert of the mentions we found a crash. It was related to the edge case where those indexes were set to NaN and then send to iOS native view. 

That didn't like it and make it crash, I think this new code brought this edge case up and so I did this fix.

cameronvoell replied:
Makes sense, thanks for the explanation!

cameronvoell reviewed (approved):
Changes here look good! 👍 

I did find one very, very small issue. I think it could be logged and saved for later, but I'll leave a video here in case you would like to address now. Basically hitting back button after adding a mention, only when there is not a space before the mention will leave the cursor in the wrong location. See gif. If there is a space before the mention (which seems like it will be the more popular case), the back button leaves the cursor in the correct spot. 
![backbutton-mention](https://user-images.githubusercontent.com/1103838/81378707-979b2500-90bc-11ea-8f1e-033505928f6a.gif)

SergioEstevao committed:
Put mentions behind the DEV flag.

SergioEstevao committed:
Merge branch 'rnmobile/stable' into rnmobile/issue_331/mention_support

SergioEstevao committed:
Merge branch 'rnmobile/release-1.28.0' into rnmobile/issue_331/mention_support

SergioEstevao committed:
Disable lint check.

SergioEstevao committed:
Merge branch 'master' into rnmobile/issue_331/mention_support

SergioEstevao committed:
Remove space.
